invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,192 @@
+import datetime
+import shutil
+import sqlite3
+import threading
+from logging import Logger
+from pathlib import Path
+from typing import Callable, Optional, TypeAlias
+
+from .sqlite_common import sqlite_memory
+
+MigrateCallback: TypeAlias = Callable[[sqlite3.Cursor], None]
+
+
+class MigrationError(Exception):
+    pass
+
+
+class Migration:
+    def __init__(
+        self,
+        version: int,
+        migrate: MigrateCallback,
+    ) -> None:
+        self.version = version
+        self.migrate = migrate
+
+
+class MigrationSet:
+    def __init__(self, table_name: str, migrations: list[Migration]) -> None:
+        self.table_name = table_name
+        self.migrations = migrations
+
+
+class SQLiteMigrator:
+    """
+    Handles SQLite database migrations.
+
+    Migrations are registered with the `register_migration_set` method. They are applied on
+    application startup with the `run_migrations` method.
+
+    A `MigrationSet` is a set of `Migration`s for a single table. Each `Migration` has a `version`
+    and `migrate` callback. The callback is provided with a `sqlite3.Cursor` and should perform the
+    any migration logic. Committing, rolling back transactions and errors are handled by the migrator.
+
+    Migrations are applied in order of version number. If the database does not have a version table
+    for a given table, it is assumed to be at version 0. The migrator creates and manages the version
+    tables.
+
+    If the database is a file, it will be backed up before migrations are applied and restored if
+    there are any errors.
+    """
+
+    def __init__(self, db_path: Path | str, lock: threading.RLock, logger: Logger):
+        self._logger = logger
+        self._conn = sqlite3.connect(db_path, check_same_thread=False)
+        self._cursor = self._conn.cursor()
+        self._lock = lock
+        self._db_path = db_path
+        self._migration_sets: set[MigrationSet] = set()
+
+    def _get_version_table_name(self, table_name: str) -> str:
+        """Returns the name of the version table for a given table."""
+        return f"{table_name}_version"
```

Comment:
Why version tables independently? In my mind, this could introduce complexity if there are interactions between tables that require migrations to be applied in a particular order.

I think its more natural to version the DB as a whole. And this is what most migration libraries that I'm aware of do (e.g. prisma, alembic).

Reply:
It's independent because:
- We have a pattern where each service manages its own tables
- I didn't know any better

Very happy to learn lessons from other libraries and do this the right way!